lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +182,212 @@ class Loader {
-    if (resolve !== undefined)
-      this._resolve = FunctionPrototypeBind(resolve, null);
-    if (dynamicInstantiate !== undefined) {
-      process.emitWarning(
-        'The dynamicInstantiate loader hook has been removed.');
-    }
-    if (getFormat !== undefined) {
-      this._getFormat = FunctionPrototypeBind(getFormat, null);
-    }
-    if (getSource !== undefined) {
-      this._getSource = FunctionPrototypeBind(getSource, null);
+    const loader = customLoader || defaultLoader;
+    const loaded = await loader(url, {}, defaultLoader);
+
+    if (typeof loaded !== 'object') {
+      throw new ERR_INVALID_RETURN_VALUE(
+        'object',
+        'loader load',
+        loaded,
+      );
     }
-    if (transformSource !== undefined) {
-      this._transformSource = FunctionPrototypeBind(transformSource, null);
+
+    const {
+      format,
+      source,
+    } = loaded;
+
+    if (format === null) {
+      const dataUrl = RegExpPrototypeExec(
+        /^data:([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/,
+        url,
+      );
+
+      throw new ERR_INVALID_MODULE_SPECIFIER(
+        url,
+        dataUrl ? `has an unsupported MIME type "${dataUrl[1]}"` : ''
+      );
     }
-    if (getGlobalPreloadCode !== undefined) {
-      this._getGlobalPreloadCode =
-        FunctionPrototypeBind(getGlobalPreloadCode, null);
+
+    if (typeof format !== 'string') {
+      throw new ERR_INVALID_RETURN_PROPERTY_VALUE(
+        'string',
+        'loader resolve',
+        'format',
+        format,
+      );
     }
+
+    if (
+      source &&
+      typeof source !== 'string' &&
+      !(source instanceof SharedArrayBuffer) &&
+      !(source instanceof Uint8Array)
+    ) throw ERR_INVALID_RETURN_PROPERTY_VALUE(
+      'string, SharedArrayBuffer, or Uint8Array',
+      'loader load',
+      'source',
+      source
+    );
+
+    return {
+      format,
+      source,
+    };
   }
 
-  runGlobalPreloadCode() {
-    if (!this._getGlobalPreloadCode) {
-      return;
-    }
-    const preloadCode = this._getGlobalPreloadCode();
-    if (preloadCode === null) {
-      return;
-    }
+  preload() {
+    const count = this.#globalPreloaders.length;
+    if (!count) return;
 
-    if (typeof preloadCode !== 'string') {
-      throw new ERR_INVALID_RETURN_VALUE(
-        'string', 'loader getGlobalPreloadCode', preloadCode);
-    }
-    const { compileFunction } = require('vm');
-    const preloadInit = compileFunction(preloadCode, ['getBuiltin'], {
-      filename: '<preload>',
-    });
-    const { NativeModule } = require('internal/bootstrap/loaders');
-
-    preloadInit.call(globalThis, (builtinName) => {
-      if (NativeModule.canBeRequiredByUsers(builtinName)) {
-        return require(builtinName);
+    for (let i = count - 1; i > -1; i--) {
+      const preloadCode = this.#globalPreloaders[i]();
+      if (preloadCode === null) return;
+
+      if (typeof preloadCode !== 'string') {
```

Comment:
OH! Good catch/idea. Do you know of a trick which is already available to get that, or would I need to persist it from `pluckHooks`?

Reply:
You probably want to store the hooks as `{ loader, hook }` pairs or something along those lines.

Reply:
The debugging information is really just the `loaderUrl` I guess? In which case perhaps `loaderUrl, hook`.

Reply:
Owww, an additional argument in favour of my proposal to chaining cuz then these can be stored as a map keyed by loaderUrl (and then `for (const [loaderUrl, hook] of this.#resolve)`); it doesn't work for the `next()` proposal because of its inception-like behaviour.